---
title: "TP NGS"
author: "Nina"
date: "12/15/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Mesorhabditis Belari genome elimination - Count matrix analysis using R

## Build the coldata (metadata)

Creation of a coldata table (text file .tsv) gathering the samples metadata: ID, name, age, timing (before/during/after elimination), elimination (yes/no), and 2 columns for differential expression analysis (see below).

```{r}
library(tidyverse)

sampleTable = read_delim("metadata.tsv") %>% as.data.frame()
rownames(sampleTable) = sampleTable$sample_ID
sampleTable
```

## Import data with tximport

```{r}
library(tximport)

dir <- "/home/rstudio/mydatalocal/tp_elimination_nc/results/quantification/"
samples <- list.dirs(dir, full.names = F) 
samples <- samples[samples != ""] # remove "." directory

samples_names <- rownames(sampleTable)

files <- file.path(dir, samples_names, "abundance.h5")
names(files) <- samples_names
files[!file.exists(files)]

txi.kallisto.tmp <- tximport(files[1], type = "kallisto", txOut = TRUE)

transcript_names <- rownames(txi.kallisto.tmp$abundance)
tx2gene <- data.frame(tx=transcript_names,
                  gene=gsub(".t[0-9]*","",transcript_names))
txi.kallisto <- tximport(files, type = "kallisto", tx2gene = tx2gene)
```
Visualization of the count matrix (columns: samples, rows: transcripts)
```{r}
txi.kallisto
```

## Differential expression analysis

### Build the DESeq2 object

Goal: build a DDS object including the normalized count matrix, coldata, and experimental design.

```{r}
library(DESeq2)

dds <- DESeqDataSetFromTximport(txi.kallisto, sampleTable, ~ elimination )
dds <- DESeq(dds)
vsd <- vst(dds, blind=FALSE)
```

### Principal component analysis

Running a PCA analysis to visualize what separates our data: we obtain a scatter plot in 2 dimensions (axes: 2 principal components that best disperse the data, with the largest variance).

```{r}
plotPCA(vsd, intgroup=c("age"))
```
As expected, we obtain:
- a close proximity between the duplicates highlighting the replicability of the results from samples of the same age
- a time curve differentiating samples according to the developmental state (axis 1: organization of the data according to the timing before/during/after elimination), evidencing that the higher data dispersion is due to developmental processes
- a separation of the "elimination" samples from the others (along the axis 2), indicating that genome elimination is linked to a differential gene expression.

### Further analysis

We want to compare gene expression during genome elimination VS before or after elimination.

The results are presented in a plot with each dot being a gene. The LogFoldChange in the ordinate axis represents the intensity of differential expression between the 2 conditions (during elimination / in absence of elimination). Dots above the LogFoldChange=0 line are the genes up-regulated during genome elimination (down-regulated genes are under the 0 line), and the blue dots have a significant p-value.

```{r}
res <- results(dds)
```

```{r}
resultsNames(dds)
```

```{r}
plotMA(res, ylim=c(-30,30))
```
Now, we want to identify these differentially expressed genes. To do so, we organize the plot results in a table.

```{r}
res_df <- as.data.frame(res) %>%rownames_to_column("gene_ID")
head(res_df)
# head: displays the first information
```

To sum up the differential gene expression that occurs during genome elimination, we can use the summary function:

```{r}
summary(res)
```

Out of 16875 genes, we obtain 1172 (6.9%) up-regulated genes and 490 (2.9%) down-regulated genes.

Then, we can try to isolate significant up- or down-regulated genes.

For instance, to isolate the most down-regulated genes with a significance threshold of 0.05, we can use the following subset function and we obtain 52 genes.

```{r}
subset(res_df, log2FoldChange < -20 & baseMean > 10 & padj < 0.05)
#padj : adjusted p-value (significative blue dots)
```

We can also isolate the most up-regulated genes with a significance threshold of 0.05, and we obtain only 1 gene.

```{r}
subset(res_df, log2FoldChange > 20 & baseMean > 10 & padj < 0.05)
```

How could we acquire more information about these differentially expressed genes?
We could, for instance, plot their expression profile and identify the protein domains associated with their transcripts.

#### Discover some analysis tools

Plot_counts function to display gene expression:

```{r}
plot_counts <- function(X) {
count_df <- counts(dds, normalized=T)
count_dfl = count_df %>% as.data.frame() %>%  rownames_to_column("gene_ID") %>% pivot_longer(-gene_ID,names_to = "sample_ID", values_to = "norm_counts")


df_tmp = subset(count_dfl, gene_ID %in% X) %>% left_join(sampleTable)

df_tmp <- df_tmp %>% mutate(age = factor(age, levels = unique(age)),
                            gene_ID = factor(gene_ID,levels=X))


ggplot(df_tmp, aes(x=age, y = norm_counts, col = age)) + theme_bw() +
  geom_point() +
  facet_wrap(~gene_ID, scales = "free_y" ) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
}
```

Function to get the transcripts annotations from WormBase:

```{r}
#BiocManager::install("biomaRt") # only one time
library(biomaRt)
mart <- useMart("parasite_mart", dataset = "wbps_gene", host = "https://parasite.wormbase.org", port = 443)

genes_annotation <- getBM(mart = mart,
                          filters = c("species_id_1010"),
                          value = list("mebelaprjeb30104"),
                          attributes = c("wbps_gene_id",
                                         "caelegprjna13758_gene_name",
                                         "caelegprjna13758_homolog_ensembl_peptide",
                                         "interpro_id",
                                         "interpro_short_description")
)
#2 first columns: homologue gene and protein names in C.elegans
#interpro: database of protein domains
#3rd and 4th columns: ID and description of protein domains
```

To visualize the results (but not all of them, because annotated genes are too numerous):

```{r}
genes_annotation[1:15,]
#1:15 indicates that we want the first 15 rows
```

#### Analysis 1

First, we're looking for genes with a significative up-regulation during genome elimination.

```{r fig.height=14, fig.width=14}
genes_list_1 = subset(res_df, padj < 0.005 & baseMean > 10 & log2FoldChange > 0)  %>% arrange (padj)
# subset: filters are applied in certain columns to get only some transcripts and we obtain the object gene_list_1
genes_list_1
plot_counts(genes_list_1$gene_ID[1:40])
# this object is showed with the plot_counts function
```
We isolate 349 significantly (p < 0.005) up-regulated different genes, sorted by increasing p-value (the first genes that are showed are the most significantly up-regulated during elimination). To further investigate them, we can search for the most significant of them on WormBase and see if they have orthologs in C.Elegans. Moreover, we can look for the effect of their know-down in C.Elegans.

Now, we want to know what kind of proteins they are coding for. Thus, we cross our results with gene annotations from WormBase to obtain protein domains associated with the gene transcripts in M.Belari (and C.elegans).

```{r fig.height=12, fig.width=12}
de_genes_with_annot_1 <- subset(genes_annotation, wbps_gene_id %in% genes_list_1$gene_ID)

de_genes_with_annot_1$wbps_gene_id <- factor(de_genes_with_annot_1$wbps_gene_id, levels = genes_list_1$gene_ID)

de_genes_with_annot_1 <- de_genes_with_annot_1%>% arrange(wbps_gene_id)
# de = differential expression
# to know what is this genes subset, we retrieve from the annotation table the genes corresponding to our subset
de_genes_with_annot_1
plot_counts(unique(de_genes_with_annot_1$wbps_gene_id)[1:40])
```

Since we're looking for proteins that could have an active role in genome elimination, we then filter the results (our genes subset) by using 4 protein domains keywords: exonuclease, transposase, helicase and DNase.

```{r}
toMatch = c("exonuclease", "transposase", "helicase", "DNase")
exo_transpo_heli_de_genes = de_genes_with_annot_1[grep(paste(toMatch,collapse="|"), de_genes_with_annot_1$interpro_short_description, ignore.case = T),]

# grep: looks for a pattern in a character chain
# |: we're looking for either an exonuclease, a transposase, or a helicase

exo_transpo_heli_de_genes

plot_counts(unique(exo_transpo_heli_de_genes$wbps_gene_id))
```
We obtain 9 genes that code for potentially interesting proteins during genome elimination. Thanks to the plot_counts function, we can see if their expression is strikingly up-regulated during elimination (4 cells and 8 cells stages) and null before and after elimination. Some of them would be particularly relevant to further investigate (e.g., knock-downing them to see if genome elimination is impaired), such as g25713, g27038 and g6879.

#### Analysis 2 : down-regulated genes

Then, we've looked for genes with a significative down-regulation during genome elimination. This analysis has been conducted by Quentin.

#### Analysis 3 : differential expression comparison between 2 other consecutive stages VS all the other stages

To separate genome elimination from developmental processes, we would like to conduct several control analyses.

First, we wonder if the differential gene expression observed during 4-8 cells stages is due to elimination or development. To do so, instead of comparing the 4-8 cells stages (when genome elimination occurs) VS all the other stages, we compare 8-8+5h cells stages (during-after elimination) and 2-4 cells stages (before-during elimination) VS all the other stages.

##### 8cells + 8cells_5h VS all

###### Build the DESeq2 object

```{r}
dds2 <- DESeqDataSetFromTximport(txi.kallisto, sampleTable, ~ de_analysis )
dds2 <- DESeq(dds2)
vsd2 <- vst(dds2, blind=FALSE)
```
###### Plot

```{r}
res2 <- results(dds2)
```

```{r}
resultsNames(dds2)
```

```{r}
plotMA(res2, ylim=c(-30,30))
```
```{r}
summary(res2)
```

The proportion of down-regulated genes is similar when comparing 4-8 cells or 8-8+5h cells to all the other stages (respectively 2.9 and 2.6%), but the proportion of up-regulated genes is lower for the comparison with 8-8+5h cells (2%) than with 4-8 cells (6.9%).

##### 2cells + 4cells VS all

###### Build the DESeq2 object

```{r}
dds3 <- DESeqDataSetFromTximport(txi.kallisto, sampleTable, ~ de_analysis)
dds3 <- DESeq(dds3)
vsd3 <- vst(dds3, blind=FALSE)
```

###### Plot

```{r}
res3 <- results(dds3)
```

```{r}
resultsNames(dds3)
```

```{r}
plotMA(res2, ylim=c(-30,30))
```

The proportion of down-regulated genes is higher for the comparison with 2-4 cells or 8-8+5h cells to all the other stages (respectively 2.9 and 2.6%), but the proportion of up-regulated genes is lower for the comparison with 8-8+5h cells (2%) than with 4-8 cells (6.9%). 